File: apps/vscode-extension/src/core/credits.ts
/** Credit accounting for Copilot usage.
 *
 *  GitHub Copilot meters usage in "AI credits" where 1 credit = $0.01 of token
 *  cost. The real per-token credit rates ship with the user's account in the
 *  Copilot Chat catalog (`models.json` → `billing.token_prices`), whose
 *  `input_price` / `output_price` values are already denominated in credits per
 *  1,000,000 tokens. We use those real rates when available and fall back to the
 *  maintained MODEL_PRICING table (USD × 100) otherwise.
 *
 *  The session logs record only the total prompt-token count (not the
 *  cached/uncached split), so prompt tokens are priced at the standard input
 *  rate — a slightly conservative approximation. */

import * as fs from 'fs/promises';
import * as path from 'path';
import { MODEL_PRICING } from '../features/costEstimator/pricing/models';
import { normalizeModelId } from '../features/costEstimator/calc/modelSelection';
import { getWorkspaceStorageRoot } from './discovery';

/** USD value of a single AI credit. */
export const AI_CREDIT_USD_VALUE = 0.01;

const TOKENS_PER_UNIT = 1_000_000;

/** Per-model credit rates, in credits per 1,000,000 tokens. */
export interface CreditRate {
  inputPerMillion: number;
  outputPerMillion: number;
}

/** Map of normalized model id → credit rate. */
export type CreditRateMap = Map<string, CreditRate>;

/** Fallback rate derived from the maintained MODEL_PRICING table (USD → credits). */
export function fallbackRateFor(modelId: string): CreditRate | undefined {
  const pricing = MODEL_PRICING[normalizeModelId(modelId)];
  if (!pricing) { return undefined; }
  return {
    inputPerMillion: pricing.inputPerMillion / AI_CREDIT_USD_VALUE,
    outputPerMillion: pricing.outputPerMillion / AI_CREDIT_USD_VALUE,
  };
}

/** Credits consumed by a single request. Returns 0 when the model is unknown. */
export function creditsForRequest(
  modelId: string | undefined,
  inputTokens: number,
  outputTokens: number,
  rates?: CreditRateMap,
): number {
  if (!modelId) { return 0; }
  const rate = rates?.get(normalizeModelId(modelId)) ?? fallbackRateFor(modelId);
  if (!rate) { return 0; }
  return (inputTokens / TOKENS_PER_UNIT) * rate.inputPerMillion
       + (outputTokens / TOKENS_PER_UNIT) * rate.outputPerMillion;
}

interface CatalogEntry {
  id?: string;
  capabilities?: { family?: string };
  billing?: {
    token_prices?: {
      default?: { input_price?: number; output_price?: number };
    };
  };
}

/** Parse a Copilot `models.json` catalog into a credit-rate map. */
export function parseCreditRatesFromCatalog(raw: string): CreditRateMap {
  const map: CreditRateMap = new Map();
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch {
    return map;
  }
  if (!Array.isArray(data)) { return map; }

  for (const entry of data as CatalogEntry[]) {
    const prices = entry?.billing?.token_prices?.default;
    if (!prices) { continue; }
    const inputPerMillion = prices.input_price;
    const outputPerMillion = prices.output_price;
    if (typeof inputPerMillion !== 'number' || typeof outputPerMillion !== 'number') { continue; }

    const rate: CreditRate = { inputPerMillion, outputPerMillion };
    // Key by both the model id and its family so resolved ids and picker ids match.
    for (const key of [entry.id, entry.capabilities?.family]) {
      if (key) { map.set(normalizeModelId(key), rate); }
    }
  }
  return map;
}

/** Locate the most recently written `models.json` catalog across all workspaces. */
async function findNewestCatalog(storageRoot: string): Promise<string | undefined> {
  let workspaceDirs: string[];
  try {
    workspaceDirs = await fs.readdir(storageRoot);
  } catch {
    return undefined;
  }

  let newestPath: string | undefined;
  let newestMtime = -Infinity;

  for (const wsDir of workspaceDirs) {
    const logsDir = path.join(storageRoot, wsDir, 'GitHub.copilot-chat', 'debug-logs');
    let sessions: string[];
    try {
      sessions = await fs.readdir(logsDir);
    } catch {
      continue;
    }
    for (const session of sessions) {
      const candidate = path.join(logsDir, session, 'models.json');
      try {
        const stat = await fs.stat(candidate);
        if (stat.isFile() && stat.mtimeMs > newestMtime) {
          newestMtime = stat.mtimeMs;
          newestPath = candidate;
        }
      } catch {
        // not present for this session
      }
    }
  }
  return newestPath;
}

/** Load real credit rates from the Copilot catalog; empty map when unavailable. */
export async function loadCreditRates(storageRoot?: string): Promise<CreditRateMap> {
  const root = storageRoot ?? getWorkspaceStorageRoot();
  const catalogPath = await findNewestCatalog(root);
  if (!catalogPath) { return new Map(); }
  try {
    const raw = await fs.readFile(catalogPath, 'utf-8');
    return parseCreditRatesFromCatalog(raw);
  } catch {
    return new Map();
  }
}

File: apps/cli/src/copilot_usage/credits.py
"""Credit accounting for Copilot usage.

Mirrors the VS Code extension's ``core/credits.ts``. GitHub Copilot meters usage
in "AI credits" where 1 credit = $0.01 of token cost. The real per-token credit
rates ship with the user's account in the Copilot Chat catalog
(``models.json`` → ``billing.token_prices``), whose ``input_price`` /
``output_price`` values are already denominated in credits per 1,000,000 tokens.
We use those real rates when available and fall back to the maintained
MODEL_PRICING table (USD ÷ credit value) otherwise.

The session logs record only the total prompt-token count (not the
cached/uncached split), so prompt tokens are priced at the standard input rate —
a slightly conservative approximation.
"""
from __future__ import annotations

import json
from pathlib import Path

from copilot_usage.config import VSCODE_STORAGE_ROOT
from copilot_usage.pricing import AI_CREDIT_USD_VALUE, MODEL_PRICING, _strip_prefix

PER_TOKEN_UNIT = 1_000_000

# (input_per_million_credits, output_per_million_credits)
CreditRate = tuple[float, float]
CreditRateMap = dict[str, CreditRate]


def fallback_rate_for(model_id: str) -> CreditRate | None:
    """Credit rate derived from the maintained MODEL_PRICING table (USD → credits)."""
    pricing = MODEL_PRICING.get(_strip_prefix(model_id or ""))
    if pricing is None:
        return None
    return (
        pricing.input_per_million / AI_CREDIT_USD_VALUE,
        pricing.output_per_million / AI_CREDIT_USD_VALUE,
    )


def credits_for_request(
    model_id: str | None,
    input_tokens: float,
    output_tokens: float,
    rates: CreditRateMap | None = None,
) -> float:
    """Credits consumed by a single request. Returns 0 when the model is unknown."""
    if not model_id:
        return 0.0
    rate: CreditRate | None = None
    if rates:
        rate = rates.get(_strip_prefix(model_id))
    if rate is None:
        rate = fallback_rate_for(model_id)
    if rate is None:
        return 0.0
    return input_tokens / PER_TOKEN_UNIT * rate[0] + output_tokens / PER_TOKEN_UNIT * rate[1]


def parse_credit_rates_from_catalog(raw: str) -> CreditRateMap:
    """Parse a Copilot ``models.json`` catalog into a credit-rate map."""
    out: CreditRateMap = {}
    try:
        data = json.loads(raw)
    except (json.JSONDecodeError, ValueError):
        return out
    if not isinstance(data, list):
        return out
    for entry in data:
        if not isinstance(entry, dict):
            continue
        billing = entry.get("billing") or {}
        token_prices = billing.get("token_prices") or {}
        prices = token_prices.get("default") or {}
        inp = prices.get("input_price")
        out_p = prices.get("output_price")
        if isinstance(inp, bool) or isinstance(out_p, bool):
            continue
        if not isinstance(inp, (int, float)) or not isinstance(out_p, (int, float)):
            continue
        rate: CreditRate = (float(inp), float(out_p))
        capabilities = entry.get("capabilities") or {}
        for key in (entry.get("id"), capabilities.get("family")):
            if isinstance(key, str) and key:
                out[_strip_prefix(key)] = rate
    return out


def _find_newest_catalog(storage_root: Path) -> Path | None:
    """Locate the most recently written ``models.json`` across all workspaces."""
    newest: Path | None = None
    newest_mtime = -1.0
    try:
        ws_dirs = list(storage_root.iterdir())
    except OSError:
        return None
    for ws_dir in ws_dirs:
        logs_dir = ws_dir / "GitHub.copilot-chat" / "debug-logs"
        if not logs_dir.is_dir():
            continue
        for candidate in logs_dir.glob("*/models.json"):
            try:
                mtime = candidate.stat().st_mtime
            except OSError:
                continue
            if mtime > newest_mtime:
                newest_mtime = mtime
                newest = candidate
    return newest


def load_credit_rates(storage_root: Path | None = None) -> CreditRateMap:
    """Load real credit rates from the Copilot catalog; empty map when unavailable."""
    root = Path(storage_root) if storage_root is not None else VSCODE_STORAGE_ROOT
    catalog = _find_newest_catalog(root)
    if catalog is None:
        return {}
    try:
        return parse_credit_rates_from_catalog(catalog.read_text(encoding="utf-8"))
    except OSError:
        return {}
